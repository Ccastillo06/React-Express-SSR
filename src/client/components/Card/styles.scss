@import '../../index.scss';

.Card {
  &__container {
    margin: 1rem;
    background: white;
    border: 1px solid lightgray;
    border-top-width: 2px;
    border-left-width: 2px;
    border-radius: 0.5rem;
    padding: 1rem;
    box-sizing: border-box;
    max-width: 300px;
    width: 100%;
    box-shadow: 4px 4px 5px 0px rgba(0, 0, 0, 0.5);
    text-align: center;

    transition: transform 200ms ease-in-out;

    &:hover {
      cursor: pointer;
      transform: scale(1.1);
    }

    &--large {
      max-width: 350px;

      &:hover {
        transform: none;
      }
    }
  }

  &__title {
    padding-right: 0.5rem;
  }

  &__picture {
    width: 100%;
    height: auto;
    border-radius: 0.5rem;
    box-shadow: 2px 2px 4px 0px rgba(0, 0, 0, 0.5);
  }

  &__info {
    border-bottom: 2px solid lightgray;
    border-radius: 0.5rem;
    padding: 0 1rem 0.5rem 1rem;
  }

  &__link-button {
    @extend .link;

    background-color: #2D3047;
    padding: 0.25rem 0.5rem;
    border-radius: 0.25rem;
    color: white;
    font-size: 0.825rem;

    &:hover {
      color: white;
      background-color: coral;
      transform: scale(1.1);
    }
  }
}
