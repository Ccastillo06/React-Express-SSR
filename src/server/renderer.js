import React from 'react';
import { renderToString } from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { renderRoutes } from 'react-router-config';
import StyleContext from 'isomorphic-style-loader/StyleContext';

import createTemplate from './createTemplate';
import createServerStore from './createServerStore';

import Routes from '../client/Routes';

export default function renderer(req) {
  const { store, storeScript } = createServerStore();

  const css = new Set(); // CSS for all rendered React components
  // eslint-disable-next-line no-underscore-dangle
  const insertCss = (...styles) => styles.forEach(style => css.add(style._getCss()));

  const content = renderToString(
    <Provider store={store}>
      <StaticRouter location={req.path}>
        <StyleContext.Provider value={{ insertCss }}>
          <div>{renderRoutes(Routes)}</div>
        </StyleContext.Provider>
      </StaticRouter>
    </Provider>,
  );

  return createTemplate(content, storeScript, css);
}
