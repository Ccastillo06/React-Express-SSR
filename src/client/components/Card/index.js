import React from 'react';
import { Link } from 'react-router-dom';
import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import withStyles from 'isomorphic-style-loader/withStyles';

import { selectCharacterById } from '../../store/namespaces/characters';

import styles from './styles.scss';

function Card({ id, noLink, isLarge }) {
  const character = useSelector(s => selectCharacterById(s, id));

  const { id: charId, name, status, image, species, gender, location } = character;

  return (
    <div className={`Card__container ${isLarge ? 'Card__container--large' : ''}`}>
      <div className="header flex-center mb-1">
        <h2 className="Card__title ellipsis">{name}</h2>
        <h3 className="regular capitalize">{status}</h3>
      </div>

      <img src={`${image}`} alt={`${name}`} className="Card__picture mb-1" />

      <div className="Card__info flex-center mb-1">
        <h3 className="regular">{species}</h3>
        <h3 className="regular">{gender}</h3>
      </div>

      <h3 className="ellipsis mb-1">{location.name}</h3>

      {!noLink ? (
        <div className="text-right">
          <Link className="Card__link-button" to={`/character/${charId}`}>
            About
          </Link>
        </div>
      ) : null}
    </div>
  );
}

Card.propTypes = {
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  noLink: PropTypes.bool,
  isLarge: PropTypes.bool,
};

Card.defaultProps = {
  noLink: false,
  isLarge: false,
};

export default withStyles(styles)(Card);
