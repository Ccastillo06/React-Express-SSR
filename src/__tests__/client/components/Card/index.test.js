import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import { StaticRouter } from 'react-router-dom';

import Card from '../../../../client/components/Card';

const mockStore = configureStore([]);

describe('Component Card renders as expected', () => {
  let store;
  let context = {};
  let component;

  beforeEach(() => {
    store = mockStore({
      characters: {
        charactersById: {
          1: {
            name: 'Saitama',
            gender: 'Male',
            status: 'Alive',
            location: {
              name: 'Earth',
            },
          },
        },
      },
    });

    component = renderer.create(
      <Provider store={store}>
        <StaticRouter location="someLocation" context={context}>
          <Card id="1" />
        </StaticRouter>
      </Provider>,
    );
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should show the character name as the title', () => {
    const cardTitle = component.root.findByProps({ className: 'Card__title ellipsis' });

    const expected = store.getState().characters.charactersById[1].name;
    expect(cardTitle.children[0]).toBe(expected);
  });
});
