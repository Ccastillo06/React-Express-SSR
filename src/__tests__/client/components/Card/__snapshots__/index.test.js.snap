// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Card renders as expected should render with given state from Redux store 1`] = `
<div
  className="Card__container "
>
  <div
    className="header flex-center mb-1"
  >
    <h2
      className="Card__title ellipsis"
    >
      Saitama
    </h2>
    <h3
      className="regular capitalize"
    >
      Alive
    </h3>
  </div>
  <img
    alt="Saitama"
    className="Card__picture mb-1"
    src="undefined"
  />
  <div
    className="Card__info flex-center mb-1"
  >
    <h3
      className="regular"
    />
    <h3
      className="regular"
    >
      Male
    </h3>
  </div>
  <h3
    className="ellipsis mb-1"
  >
    Earth
  </h3>
  <div
    className="text-right"
  >
    <a
      className="Card__link-button"
      href="/character/undefined"
      onClick={[Function]}
    >
      About
    </a>
  </div>
</div>
`;
